src/inference/dreamSample.ad.js

```diff
@@ -0,0 +1,138 @@
+// Coroutine of the forward pass in dream training. Samples from the guide
+// outside of mapData and from the target inside of mapData.
+// Records the traces and observations.
+
+'use strict';
+'use ad';
+
+var _ = require('underscore');
+var util = require('../util');
+var ad = require('../ad');
+var Trace = require('../trace');
+var guide = require('../guide');
+
+// TODO - should merge it back with forwardSample.js or keep it seperated?
+// There are similarities in the general structure but ultimately also
+// signficant differences and so maybe it would be better to keep these
+// as two simlpe seperated modules rather than one merged..
+module.exports = function(env) {
+
+  function DreamSample(s, k, a, wpplFn, options) {
+    this.opts = util.mergeDefaults(options, {
+      samples: 1,
+      verbose: true,
+      params: {}
+    });
+
+    if (this.opts.samples <= 0) {
+      throw 'Invalid samples amount.';
+    }
+
+    this.params = this.opts.params;
+    this.wpplFn = wpplFn;
+    this.s = s;
+    this.k = k;
+    this.a = a;
+
+    // true when inside mapData (i.e. local model)
+    // false otherwise (i.e. global model)
+    this.insideMapData = true;
+
+    this.records = [];
+    this.currRecord;
```

Comment:
Maybe it's because I'm used to work more with other languages, but it personally seems nice to me that variables that are used throughout a module are declared in the beginning, even tough it practically does nothing here, but on the other hand it's non-js style so maybe I should change that. What do you think?